crates/types/src/locality/node_location.rs

```diff
@@ -0,0 +1,539 @@
+        assert_that!(location.label_at(NodeLocationScope::Zone), eq(""));
+        assert_that!(location.label_at(NodeLocationScope::Node), eq(""));
+        assert_that!(location.num_defined_scopes(), eq(1));
+        assert_that!(
+            location.domain_string(NodeLocationScope::Region, None),
+            eq("us-east2")
+        );
+        assert_that!(
+            location.domain_string(NodeLocationScope::Zone, None),
+            eq("us-east2")
+        );
+        assert_that!(
+            location.domain_string(NodeLocationScope::Node, None),
+            eq("us-east2")
+        );
+        assert_that!(
+            location.domain_string(NodeLocationScope::Node, Some(PlainNodeId::new(4))),
+            eq("us-east2:N4")
+        );
+
+        assert_that!(location.to_string(), eq("us-east2"));
+        assert_that!(location, eq(NodeLocation::from_str("us-east2").unwrap()));
+    }
+
+    #[test]
+    fn node_location_parsing_with_empty_labels() {
+        // valid case 1 -- empty str
+        let location = NodeLocation::from_str("").unwrap();
+        assert_that!(
+            location.smallest_defined_scope(),
+            eq(NodeLocationScope::Root)
+        );
+        assert_that!(location.label_at(NodeLocationScope::Region), eq(""));
+        assert_that!(location.label_at(NodeLocationScope::Zone), eq(""));
+        assert_that!(location.label_at(NodeLocationScope::Node), eq(""));
+        assert_that!(location.num_defined_scopes(), eq(0));
+        assert_that!(
+            location.domain_string(NodeLocationScope::Region, None),
+            eq("")
+        );
+        assert_that!(
+            location.domain_string(NodeLocationScope::Node, None),
+            eq("")
+        );
+        assert_that!(
+            location.domain_string(NodeLocationScope::Node, Some(5.into())),
+            eq(":N5")
+        );
+
+        assert_that!(location, eq(NodeLocation::from_str(".").unwrap()));
+        assert_that!(location, eq(NodeLocation::from_str(" ").unwrap()));
+        assert_that!(location, eq(NodeLocation::from_str("   ").unwrap()));
+
+        // valid case 1 -- region only
+        let location = NodeLocation::from_str("us-east1").unwrap();
+        assert_that!(location.label_at(NodeLocationScope::Region), eq("us-east1"));
+        assert_that!(location.label_at(NodeLocationScope::Zone), eq(""));
+        assert_that!(location.label_at(NodeLocationScope::Node), eq(""));
+        assert_that!(location.num_defined_scopes(), eq(1));
+        assert_that!(
+            location.domain_string(NodeLocationScope::Region, None),
+            eq("us-east1")
+        );
+        assert_that!(
+            location.domain_string(NodeLocationScope::Node, None),
+            eq("us-east1")
+        );
+        assert_that!(
+            location.domain_string(NodeLocationScope::Node, Some(5.into())),
+            eq("us-east1:N5")
+        );
```

Comment:
Maybe create a separate test function since the labels are no longer empty here.

Reply:
Agreed, moved it to the simple parsing test.